File: day11/haskell/day11.hs
import Control.Arrow ((&&&))

type Coord = (Int, Int, Int)

move :: Coord -> String -> Coord
move (x, y, z) "n"  = (x - 1, y + 1, z)
move (x, y, z) "ne" = (x, y + 1, z - 1)
move (x, y, z) "nw" = (x - 1, y, z + 1)
move (x, y, z) "s"  = (x + 1, y - 1, z)
move (x, y, z) "se" = (x + 1, y, z - 1)
move (x, y, z) "sw" = (x, y - 1, z + 1)

parse :: String -> [String]
parse = words . map f
    where f ',' = ' '
          f x   = x

solve :: [String] -> (Int, Int)
solve = (last &&& maximum) . map f . scanl move (0, 0, 0)
    where f (x, y, z) = maximum [abs x, abs y, abs z]
        
main :: IO ()
main = do
    input <- parse <$> readFile "day11/input.txt"
    print . solve $ input

File: day11/typescript/day11.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

type Coord = [number, number, number];
type Moves = { [dir: string]: (coord: Coord) => Coord };

const moves: Moves = {
    n: ([x, y, z]) => [x - 1, y + 1, z],
    ne: ([x, y, z]) => [x, y + 1, z - 1],
    nw: ([x, y, z]) => [x - 1, y, z + 1],
    s: ([x, y, z]) => [x + 1, y - 1, z],
    se: ([x, y, z]) => [x + 1, y, z - 1],
    sw: ([x, y, z]) => [x, y - 1, z + 1],
};

const path = (dirs: string[]) => R.scan<string, Coord>(
    (acc, curr) => moves[curr](acc),
    [0, 0, 0],
    dirs,
);

const part1 = (dirs: string[]) => {
    const end = R.last(path(dirs))!;
    return Math.max(...end.map(Math.abs));
};

const part2 = (dirs: string[]) => {
    const points = path(dirs);
    return Math.max(...R.map(p => Math.max(...p.map(Math.abs)), points));
};

(async () => {
    const input = await promisify(readFile)('day11/input.txt', 'utf8');
    const dirs = input.split(',');

    console.log(part1(dirs));
    console.log(part2(dirs));
})();
